api/v1alpha1/gateway_types.go

```diff
@@ -238,6 +238,39 @@ type LocalObjectReference struct {
 	Name string `json:"name" protobuf:"bytes,3,opt,name=name"`
 }
 
+// RouteBindingSelector defines a schema for associating routes with the Gateway.
+// No routes will be associated to the Gateway if NamespaceSelector and RouteSelector
+// are undefined. If both NamespaceSelector and RouteSelector are defined, only
+// routes matching the RouteSelector within the NamespaceSelector namespaces are
+// associated with the Gateway.
+type RouteBindingSelector struct {
+	// NamespaceSelector specifies a set of namespace labels used for selecting
+	// routes to associate with the Gateway. If NamespaceSelector is defined,
+	// all routes within the NamespaceSelector namespaces are associated with
+	// the Gateway. An empty NamespaceSelector matches all namespaces.
+	//
+	// If undefined, namespace labels are not used for associating routes to
+	// the gateway.
+	//
+	// Support: Core
+	//
+	// +optional
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector" protobuf:"bytes,1,opt,name=namespaceSelector"`
```

Comment:
I had a comment in the security doc that I'll re-iterate here: https://docs.google.com/document/d/14-mDATLLiF70wFaUPDXxFjO9BDrl9863Lk8um2RaTLs/edit

The labelSelector seems to be against the original motivation for adding namespace based selection in the first place.

As far as I understand, the original motivation for adding namespaceSelector was to make app devs ask for an explicit permission from the cluster operator before a route from a namespace that the app dev controls can be attached to a Gateway.

If instead a labelSelector is added, now the association of Gateway and Route is at the mercy of whoever has write access to the namespaces. Doesn't it defeat the purpose of namespaceSelector itself?

Reply:
@hbagdi here's the use case behind `RouteSelector`:
- I have a single namespace that is shared among multiple dev teams containing a ton of routes.
- I have multiple gateways.
- I need the ability to associate different routes from the shared namespace to different gateways.

Thoughts on how to achieve this use case?

Reply:
I'm all in for label selector for routes. I'm trying to understand the reasoning behind label selector for namespaces.